File: src/examples/SUPERball/AppSUPERSend.cpp
#include <stdlib.h>
#include <string>
#include <iostream>
#include <sstream>

#include <zmq.hpp>
#include <unistd.h>

#include <math.h>

int main() {
    zmq::context_t context(1);
    zmq::socket_t socket(context, ZMQ_REQ);
    std::cout << "Connection to SUPERball server" << std::endl;

    socket.connect("tcp://localhost:5555");

    //do an infinite number of requests, just making the ball get bigger and smaller

    int num_actuators = 25;

    double lifetime = 0.0;
    
    for(;;) {
        //compute the behaviour

        double tr = sin(2*M_PI*((double)1/4)*lifetime)+1.0;

        std::stringbuf buffer;
        std::ostream msg (&buffer);

        for (int i = 0; i < num_actuators; i++) {
            msg << 1*tr << ",";
        }

        std::string msg_str(buffer.str());
        buffer.str(""); //reset the buffer
        zmq::message_t req(msg_str.length());
        memcpy(req.data(), &msg_str[0], msg_str.length());
        std::cout << "sending request " << (char*)req.data() << std::endl;
        msg_str = "";

        socket.send(req);

        std::cout << "waiting for response (simulator time)" << std::endl;
        zmq::message_t reply;
        socket.recv(&reply);

        std::string reply_message(static_cast<char*>(reply.data()), reply.size());
        std::istringstream reply_ss(reply_message);
        reply_ss >> lifetime;

        std::cout << "Lifetime:" << lifetime << std::endl;

        usleep(1000);
    }

}

File: src/examples/SUPERball/AppSUPERSend.py
#!/usr/bin/python3

import zmq
import time
import numpy as np
import math

def main():
    port = 5555 

    context = zmq.Context()
    socket = context.socket(zmq.REQ)
    socket.connect("tcp://localhost:%s" % port)

    lifetime = 0.0

#     for _ in range(200):
    while True:
        tr = (math.sin(2*math.pi*(0.2)*lifetime)+1.0) / 2

        msg = ','.join([str(tr * (math.sin(2*math.pi*i/25)+1.0)/2) for i in range(25)])
        print(msg)
        socket.send_string(msg)
        message = socket.recv()
        
        lifetime = float(message.split()[0])

        #lifetime = float(message)
        print(lifetime)
        


### end main    

if __name__ == '__main__':
    main()
